10-min shippable package automation (#26098)
Azure/azure-sdk-for-net: doc/Data Plane Code Generation/AzureSDKCodeGeneration_DataPlane_Quickstart.md

**Learn more**: to understand more about Azure SDK Code Geneneration, see the [Azure SDK Code Geneneration docs](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core/samples/ProtocolMethods.md).

This tutorial has following sections:

- [Azure SDK Code Geneneration Quickstart Tutorial (Data Plane)](#azure-sdk-code-geneneration-quickstart-tutorial-data-plane)
  - [Prerequisites](#prerequisites)
  - [Setup your repo](#setup-your-repo)
  - [Getting started - shippable package](#getting-started---shippable-package)
    - [Create the project folder](#create-the-project-folder)
      - [1. install dotnet tempate](#1-install-dotnet-tempate)
      - [2. dotnet new project](#2-dotnet-new-project)
      - [3. update the solution file if needed](#3-update-the-solution-file-if-needed)
    - [Generate client library](#generate-client-library)
    - [Export Public API](#export-public-api)
  - [Improve the SDK library package](#improve-the-sdk-library-package)
    - [Tests](#tests)
    - [Samples](#samples)
    - [Snippets](#snippets)
    - [README.md](#readmemd)
    - [Changelog](#changelog)

<!-- /TOC -->

> ## Prerequisites ##

Reviewer: We should consider linking the prerequistes we have in our contributing guide so we don't have multiple sets to maintain.

- Install VS 2020 (Community or higher) and make sure you have the [latest updates](https://www.visualstudio.com/).
  - Need at least .NET Framework 4.6.1 and 4.7 development tools
  - Install the **.NET Core cross-platform development** workloads in VisualStudio
- Install **.NET 6.0 SDK** for your specific platform. (or a higher version)
- Install **.NET core 3.1** for your specific platform.
- Install the latest version of [git](https://git-scm.com/downloads)
- Install [PowerShell](https://docs.microsoft.com/powershell/scripting/install/installing-powershell), version 7 or higher.
- Install [NodeJS](https://nodejs.org/) (14.x.x).

> ## Setup your repo ##

Reviewer: Why the `##` at the end of the lines?

Author: the md rule to define it as the title of chapter

Author: The ## at the end is not needed. removed.

- Fork and clone an [azure-sdk-for-net](https://github.com/Azure/azure-sdk-for-net) repo. Follow the instructions in the [.NET Contributing.md](https://github.com/Azure/azure-sdk-for-net/issues/12903) to fork and clone the `azure-sdk-for-net` repo.
- Create a branch to work in. 

## Getting started - shippable package ##

> For this tutorial, we'll create a getting started project in a branch of your fork of `azure-sdk-for-net` repo. We will use dotnet project tempalte [Azure.ServiceTemplate.Template](https://github.com/Azure/azure-sdk-for-net/) to automatically create the project.

Reviewer: ```suggestion
For this guide, we'll create a getting started project in a branch of your fork of `azure-sdk-for-net` repo. We will use dotnet project template [Azure.ServiceTemplate.Template](https://github.com/Azure/azure-sdk-for-net/) to automatically create the project.
```

Author: updated

> You can run `eng\automation\Invoke-DataPlaneGenerateShippablePackage.ps1` to generate the first-verion shippable SDK client library package directly as following

Reviewer: ```suggestion
You can run `eng\automation\Invoke-DataPlaneGenerateShippablePackage.ps1` to generate the first version shippable SDK client library package directly as following
```

Author: updated

```
eng/automation/Invoke-DataPlaneGenerateShippablePackage.ps1 -service <servicename> -namespace Azure.<group>.<source> -sdkPath <sdkrepoRootPath> -inputfile <inputfilelink> -securityScope <securityScope>
```

> e.g.

Reviewer: These should be in code blocks for easier reading as MD.

Author: updated.

pwsh /home/azure-sdk-for-net/eng/automation/Invoke-DataPlaneGenerateShippablePackage.ps1 -service sample -namespace Azure.Template.Sample -sdkPath /home/azure-sdk-for-net -inputfile https://github.com/Azure/azure-rest-api-specs/blob/73a0fa453a93bdbe8885f87b9e4e9fef4f0452d0/specification/webpubsub/data-plane/WebPubSub/stable/2021-10-01/webpubsub.json -securityScope https://sample/.default

**Note**: Use one of the following pre-approved namespace groups (https://azure.github.io/azure-sdk/registered_namespaces.html): Azure.AI, Azure.Analytics, Azure.Communication, Azure.Data, Azure.DigitalTwins, Azure.IoT, Azure.Learn, Azure.Media, Azure.Management, Azure.Messaging, Azure.ResourceManager, Azure.Search, Azure.Security, Azure.Storage, Azure.Template, Azure.Identity, Microsoft.Extensions.Azure
**Note**: namespace is the shipped package name, it should be Azure.<group>.<service>
**Note**: inputfile is the api definition file. It can be local file e.g. ./swagger/compute.json or premlink, e.g. https://github.com/dpokluda/azure-rest-api-specs/blob/be397aa65510bd4e8f87da539af2b0025f6f44ca/specification/deviceupdate/data-plane/Microsoft.DeviceUpdate/preview/2020-09-01/deviceupdate.json".

> Or you can do **step-by-step** as following.

Reviewer: Is there any reasons folks would want to do all the steps manually? It might just cause confusion having the script and the manual steps.

Author: change the following as the instruction of the script.

# Azure SDK Code Geneneration Quickstart Tutorial (Data Plane)

> The purpose of the Azure SDK is to provide a unified developer experience across Azure services. In this tutorial, we will walk through creating a new Azure SDK Generated Client library for a data plane Azure service.  The output library will have a standardized API based on the  [.NET Azure SDK Design Guidelines](https://azure.github.io/azure-sdk/dotnet_introduction.html).

Reviewer 2: We should explain in the introduction what we mean by shippable and link to further information on the process and requirements for shipping if we choose to keep using that term.

Reviewer 2: Moving to the new terminology, we no longer need this definition in the introduction.

We will use the swagger file which contains a REST API definition based on [Azure REST API Guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). We will use the swagger to generate SDK using [autorest.csharp](https://github.com/Azure/autorest.csharp).

**Learn more**: to understand more about Azure SDK Code Geneneration, see the [Azure SDK Code Geneneration docs](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core/samples/ProtocolMethods.md).

This tutorial has following sections:

- [Azure SDK Code Geneneration Quickstart Tutorial (Data Plane)](#azure-sdk-code-geneneration-quickstart-tutorial-data-plane)
  - [Prerequisites](#prerequisites)
  - [Setup your repo](#setup-your-repo)

>   - [Getting started - shippable package](#getting-started---shippable-package)

Reviewer 2: I'm a little concerned about the idea that we're calling a package without tests shippable.  Have we given any thought to this?  We also require samples according to [the guidelines](https://azure.github.io/azure-sdk/dotnet_introduction.html#dotnet-samples)

Reviewer 2: Per offline discussion, let's rename these sections to:
- “Getting started - shippable package” -> “Create starting package”
- “Improve the SDK library package” -> “Add package ship requirements”

Author: updated

    - [Create the project folder](#create-the-project-folder)

>       - [1. install dotnet tempate](#1-install-dotnet-tempate)

Reviewer 2: Nit: run a spell-checker?

```suggestion
      - [1. install dotnet template](#1-install-dotnet-template)
```

      - [2. dotnet new project](#2-dotnet-new-project)
      - [3. update the solution file if needed](#3-update-the-solution-file-if-needed)
    - [Generate client library](#generate-client-library)

>     - [Export Public API](#export-public-api)

Reviewer 3: nit fix: Public -> public(follow same format)

>   - [Improve the SDK library package](#improve-the-sdk-library-package)

Reviewer 2: Improve here sounds optional.  We should clarify that these are requirements for shipping.

Author: updated

    - [Tests](#tests)
    - [Samples](#samples)
    - [Snippets](#snippets)

>     - [README.md](#readmemd)

Reviewer 3: Why do we have `.md` only for readme? can we use the same format for all?

    - [Changelog](#changelog)

<!-- /TOC -->

## Prerequisites

> - Install VS 2020 (Community or higher) and make sure you have the [latest updates](https://www.visualstudio.com/).

Reviewer 4: From https://dotnet.microsoft.com/en-us/download/dotnet-framework,  .NET framework 4.8 is the recommended version. so do we support that version or do we only support 4.6 & 4.7

Author: yes. we will support .Net framework 4.6 and 4.7

  - Need at least .NET Framework 4.6.1 and 4.7 development tools
  - Install the **.NET Core cross-platform development** workloads in VisualStudio
- Install **.NET 6.0 SDK** for your specific platform. (or a higher version)
- Install **.NET core 3.1** for your specific platform.
- Install the latest version of [git](https://git-scm.com/downloads)
- Install [PowerShell](https://docs.microsoft.com/powershell/scripting/install/installing-powershell), version 7 or higher.
- Install [NodeJS](https://nodejs.org/) (14.x.x).

## Setup your repo

- Fork and clone an [azure-sdk-for-net](https://github.com/Azure/azure-sdk-for-net) repo. Follow the instructions in the [.NET Contributing.md](https://github.com/Azure/azure-sdk-for-net/issues/12903) to fork and clone the `azure-sdk-for-net` repo.
- Create a branch to work in. 

> ## Getting started - shippable package

Reviewer 2: Could you help clarify the goal of adding this to this section header?  Are we saying that once the reader has completed the Getting Started section, they will have a shippable package?  If so, we should probably not call it getting started, maybe "Create a shippable package" (but I'd like us to clarify what we mean by shippable as noted elsewhere).

Reviewer 2: Resolved per above.

For this guide, we'll create a getting started project in a branch of your fork of `azure-sdk-for-net` repo. We will use dotnet project template [Azure.ServiceTemplate.Template](https://github.com/Azure/azure-sdk-for-net/) to automatically create the project.

> You can run `eng\automation\Invoke-DataPlaneGenerateShippablePackage.ps1` to generate the first-version shippable SDK client library package directly as following

Reviewer 2: Per comment above, let's call this "starting package" rather than "shippable package".  Let's also define what we mean by that in this section to make it clear to readers, and per @lmazuel, let's make sure we include that it is a package that successfully passes CI in the definition.  Thanks!

```

> eng/automation/Invoke-DataPlaneGenerateShippablePackage.ps1 -service <servicename> -namespace Azure.<group>.<source> -sdkPath <sdkrepoRootPath> -inputfile <inputfilelink> -securityScope <securityScope>

Reviewer 3: It takes only AADToken as you mentioned `-securityScope`? do we also have any attribute for `AzureKeyCredential` (one example you can check [here](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/webpubsub/Azure.Messaging.WebPubSub/src/autorest.md))? if yes, can we add an example for that as well?

Author: supported

**Note**: Use one of the following pre-approved namespace groups (https://azure.github.io/azure-sdk/registered_namespaces.html): Azure.AI, Azure.Analytics, Azure.Communication, Azure.Data, Azure.DigitalTwins, Azure.IoT, Azure.Learn, Azure.Media, Azure.Management, Azure.Messaging, Azure.ResourceManager, Azure.Search, Azure.Security, Azure.Storage, Azure.Template, Azure.Identity, Microsoft.Extensions.Azure
**Note**: namespace is the shipped package name, it should be Azure.<group>.<service>
**Note**: inputfile is the api definition file. It can be local file e.g. ./swagger/compute.json or premlink, e.g. https://github.com/dpokluda/azure-rest-api-specs/blob/be397aa65510bd4e8f87da539af2b0025f6f44ca/specification/deviceupdate/data-plane/Microsoft.DeviceUpdate/preview/2020-09-01/deviceupdate.json".

> **Learn more:** the [autorest.csharp README](https://github.com/Azure/autorest.csharp#setup) has great samples showing how to add convenience APIs in the generated code. Explore this further as you design APIs for your own service.

Reviewer 3: Can we add back this section somewhere? it talks about where to add configuration and how to add convenience layer.

> #### 1. install dotnet tempate

Reviewer 3: ```suggestion
#### 1. install dotnet template
```

Author: updated

> navigate to the sdk repo root director, and run the following commands:

Reviewer 2: It looks like this comment was resolved without addressing?

> navigate to the sdk repo root director, and run the following commands:

Author: sorry. forget to push. Now updated

> dotnet sln add tests\Azure.<group>.<service>.Tests.csproj

Reviewer 3: Add new lines. It is not adding `dotnet sln.....` in new lines. Check [here](https://github.com/Azure/azure-sdk-for-net/blob/bb0fbccfc33dd27d1ec6f0870022824d47181e61/doc/Data%20Plane%20Code%20Generation/AzureSDKCodeGeneration_DataPlane_Quickstart.md#3-update-the-solution-file-if-needed).

> dotnet sln remove src\Azure.<grouop>.<service>.csproj
> dotnet sln add src\Azure.<grouop>.<service>.csproj
> dotnet sln remove tests\Azure.<grouop>.<service>.Tests.csproj
> dotnet sln add tests\Azure.<grouop>.<service>.Tests.csproj

Reviewer 3: ```suggestion
dotnet sln remove src\Azure.<group>.<service>.csproj
dotnet sln add src\Azure.<group>.<service>.csproj
dotnet sln remove tests\Azure.<group>.<service>.Tests.csproj
dotnet sln add tests\Azure.<group>.<service>.Tests.csproj
```

Author: updated

> You can continuously work on the first-version shippable package created above to refine, such as add Tests, Samples and refine the Readme.md and Changlog.

Reviewer 3: ```suggestion
You can continuously work on the first-version shippable package created above to refine, such as add Tests, Samples and refine the Readme.md and Changelog.
```

> ### Tests ###

Reviewer 3: Why is `###` needed after the Tests?

Reviewer 2: Let's please address Shivangi's comment.

Author: updated

> **Note**: Before running live tests you will need to create live test resources, please refer [this](https://github.com/Azure/azure-sdk-for-net/blob/main/eng/common/TestResources/README.md) to learn more about how to manager resources and update test environment.

Reviewer 3: It is not adding `**Note**:` in new line. Check [here](https://github.com/Azure/azure-sdk-for-net/blob/bb0fbccfc33dd27d1ec6f0870022824d47181e61/doc/Data%20Plane%20Code%20Generation/AzureSDKCodeGeneration_DataPlane_Quickstart.md#tests).

> **Note**: `Azure.<group>.<service>` is the Azure SDK package name and `<client-name>` is a client name, C# generator will generate a client which you can find in `Azure.<group>.<service>/Generated` directory.

Reviewer 3: Can we add bullet points and merge both the `**Notes**`?

# Azure SDK Code Geneneration Quickstart Tutorial (Data Plane)
The purpose of the Azure SDK is to provide a unified developer experience across Azure services. In this tutorial, we will walk through creating a new Azure SDK Generated Client library for a data plane Azure service.  The output library will have a standardized API based on the  [.NET Azure SDK Design Guidelines](https://azure.github.io/azure-sdk/dotnet_introduction.html).

We will use the swagger file which contains a REST API definition based on [Azure REST API Guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). We will use the swagger to generate SDK using [autorest.csharp](https://github.com/Azure/autorest.csharp).

**Learn more**: to understand more about Azure SDK Code Geneneration, see the [Azure SDK Code Geneneration docs](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core/samples/ProtocolMethods.md).

This tutorial has following sections:

- [Azure SDK Code Geneneration Quickstart Tutorial (Data Plane)](#azure-sdk-code-geneneration-quickstart-tutorial-data-plane)
  - [Prerequisites](#prerequisites)
  - [Setup your repo](#setup-your-repo)

>   - [Create starting package](#create-starting-package)

Reviewer 2: Would you be willing to change this to "starter package"?

Author: updated

    - [Create the project folder](#create-the-project-folder)
      - [1. install dotnet template](#1-install-dotnet-template)
      - [2. dotnet new project](#2-dotnet-new-project)
      - [3. update the solution file if needed](#3-update-the-solution-file-if-needed)
    - [Generate client library](#generate-client-library)
    - [Export public API](#export-public-api)
  - [Add package ship requirements](#add-package-ship-requirements)
    - [Tests](#tests)
    - [Samples](#samples)
    - [Snippets](#snippets)
    - [README](#readme)
    - [Changelog](#changelog)
    - [Customize](#customize)

<!-- /TOC -->

## Prerequisites

- Install VS 2020 (Community or higher) and make sure you have the [latest updates](https://www.visualstudio.com/).
  - Need at least .NET Framework 4.6.1 and 4.7 development tools
  - Install the **.NET Core cross-platform development** workloads in VisualStudio
- Install **.NET 6.0 SDK** for your specific platform. (or a higher version)
- Install **.NET core 3.1** for your specific platform.

> - Install the latest version of [git](https://git-scm.com/downloads)

Author: I think we only use git common functions (clone, checkout, commit), so it is not to require special version of git.

- Install [PowerShell](https://docs.microsoft.com/powershell/scripting/install/installing-powershell), version 7 or higher.

> - Install [NodeJS](https://nodejs.org/) (14.x.x).

Author: in generation script, we use npx autorest ..., so we will not need to install autorest, npm will download autorest dynamically

## Setup your repo

- Fork and clone an [azure-sdk-for-net](https://github.com/Azure/azure-sdk-for-net) repo. Follow the instructions in the [.NET Contributing.md](https://github.com/Azure/azure-sdk-for-net/issues/12903) to fork and clone the `azure-sdk-for-net` repo.
- Create a branch to work in. 

> 

Reviewer 5: I think we need a section here that explains some of the basic terminology, specifically `service name` and `package name`. These terms should be defined consistently with the "sdk-structure" description at

https://github.com/Azure/azure-sdk/blob/main/docs/policies/repostructure.md#sdk-directory-layout

This might also be the right place to put the information about namespaces (move from below).

Author: I will address this

## Create starting package

For this guide, we'll create a getting started project in a branch of your fork of `azure-sdk-for-net` repo. We will use dotnet project template [Azure.ServiceTemplate.Template](https://github.com/Azure/azure-sdk-for-net/) to automatically create the project.

You can run `eng\scripts\automation\Invoke-DataPlaneGenerateSDKPackage.ps1` to generate the starting SDK client library package directly as following:
```
eng/scripts/automation/Invoke-DataPlaneGenerateSDKPackage.ps1 -service <servicename> -namespace Azure.<group>.<source> -sdkPath <sdkrepoRootPath> -inputfiles <inputfilelink> -securityScope <securityScope>
```

e.g.
```
pwsh /home/azure-sdk-for-net/eng/scripts/automation/Invoke-DataPlaneGenerateSDKPackage.ps1 -service sample -namespace Azure.Template.Sample -sdkPath /home/azure-sdk-for-net -inputfiles https://github.com/Azure/azure-rest-api-specs/blob/73a0fa453a93bdbe8885f87b9e4e9fef4f0452d0/specification/webpubsub/data-plane/WebPubSub/stable/2021-10-01/webpubsub.json -securityScope https://sample/.default
```

**Note**: 

> - Use one of the following pre-approved namespace groups (https://azure.github.io/azure-sdk/registered_namespaces.html): Azure.AI, Azure.Analytics, Azure.Communication, Azure.Data, Azure.DigitalTwins, Azure.IoT, Azure.Learn, Azure.Media, Azure.Management, Azure.Messaging, Azure.ResourceManager, Azure.Search, Azure.Security, Azure.Storage, Azure.Template, Azure.Identity, Microsoft.Extensions.Azure

Reviewer 2: Do we need the namespaces to be listed here? How will we keep it in sync with what's at https://azure.github.io/azure-sdk/registered_namespaces.html?

Author: In the webside, it just list the registered namespaces, not the approved groups. Currently we have no webside to list approved groups, so I list in the doc. If we have website for groups, I will update the doc with the link,

Reviewer 2: The groups are listed here: https://azure.github.io/azure-sdk/dotnet_introduction.html#dotnet-namespaces-approved-list

Author: add the approved groups websit link in the doc

> - namespace is the shipped package name, it should be Azure.<group>.<service>

Reviewer 2: ```suggestion
- namespace is the shipped package name, it should be `Azure.<group>.<service>`
```

(Without escaping, the `<..>` are hidden in the .md file.

Author: updated

- inputfiles is the api definition files, separated by semicolon if more than one. The api definition file can be local file e.g. ./swagger/compute.json or premlink, e.g. https://github.com/Azure/azure-rest-api-specs/blob/73a0fa453a93bdbe8885f87b9e4e9fef4f0452d0/specification/webpubsub/data-plane/WebPubSub/stable/2021-10-01/webpubsub.json.

> - Both AADToken and AzureKey authentication are supported. If your service support AADToken, just set the parameter **securityScope**(the security scope), and if your service support AzureKey authentication, set parameter **securityHeaderName**( the security header name). You also can provide both if your service support two authentications.

Reviewer 2: Would it be possible to link to more information regarding what `AADToken` and `AzureKey` signify?

Author: I couldn't find the information link for AADToken and AzureKey.

Reviewer 2: @christothes do you know where we might find good docs/explanation on when a service partner would use one or the other?

Reviewer 6: I don't know of any specific docs, but generally you would provide all of them that are supported by the service. AzureKey auth is specific to each service, but generally it is documented by that service. Examples include Azure Storage - https://docs.microsoft.com/en-us/azure/storage/blobs/authorize-data-operations-portal#use-the-account-access-key

AADToken is for any service that supports TokenCredential auth https://docs.microsoft.com/en-us/dotnet/api/azure.core.tokencredential?view=azure-dotnet

Reviewer 2: Thanks, @christothes!  @chunyu3, could we include Chris's explanations in this section to help readers understand?

Author: added

The script `eng\scripts\automation\Invoke-DataPlaneGenerateSDKPackage.ps1` will do **step-by-step** as following.

> ### Create the project folder

Reviewer 2: The way these headers end up in the table of contents, it seems like we're asking the reader to do this, rather than telling them the script will do this.  Would it be possible to adjust the formatting, so we know we're in the context of a script and that these are not actions the reader will take?

Reviewer 2: Curious: would it make sense to link to the script, and then summarize the steps in words, rather than showing the commands that are in the script?

Author: update to use list for steps of the Invoke-DataPlaneGenerateSDKPackage.ps1 to avoid confusing

Create a SDK library project, configuration file, or solution based on the specified template.
#### 1. install dotnet template
navigate to the sdk repo root director, and run the following commands:

```

> dotnet new -i sdk/template-dpg/Azure.ServiceTemplate.Template

Reviewer 2: nit: would it make sense to write out `install` so it's more easily human-readable?

```suggestion
dotnet new --install sdk/template-dpg/Azure.ServiceTemplate.Template
```

Author: updated

```

#### 2. dotnet new project

> - create project folder 'Azure.<group>.<service>'. e.g. Azure.IoT.DeviceUpdate under sdk/<service> folder

Reviewer 2: ```suggestion
- create project folder `Azure.<group>.<service>`. e.g. Azure.IoT.DeviceUpdate under `sdk/<service>` folder
```

Otherwise, the md looks like this:
![image](https://user-images.githubusercontent.com/21269347/149437962-d66e582a-3fcc-4b6c-9a6d-e342ff0c6739.png)

- new project: navigate to the project folder, and run 'dotnet new' as following:
  
```
sdk\<your-service-name>\Azure.<group>.<service>> dotnet new dataplane --libraryName [Client-Library-Title] --swagger [input-swagger-file-path] --securityScopes [security-scopes] --force
```
e.g.
```
dotnet new dataplane --libraryName DeviceUpdate --swagger https://github.com/Azure/azure-rest-api-specs/blob/23dc68e5b20a0e49dd3443a4ab177d9f2fcc4c2b/specification/deviceupdate/data-plane/Microsoft.DeviceUpdate/preview/2021-06-01-preview/deviceupdate.json --securityScopes https://api.adu.microsoft.com/.default --force
```

#### 3. update the solution file if needed
Run 'dotnet sln' as following to update the projects in the solution file:

```
dotnet sln remove src\Azure.<group>.<service>.csproj
dotnet sln add src\Azure.<group>.<service>.csproj
dotnet sln remove tests\Azure.<group>.<service>.Tests.csproj
dotnet sln add tests\Azure.<group>.<service>.Tests.csproj
```

**Learn more:** to understand more about the Azure SDK repo structure, see [Repo Structure](https://github.com/Azure/azure-sdk/blob/master/docs/policies/repostructure.md) in the `azure-sdk` repo.

> ### Generate client library

Reviewer 2: This is a step we want the reader to take, whereas **"Create the project folder"** was something the script does for them.  Would it be possible to update the formatting so this is more clear?

Author: The script Invoke-DataPlaneGenerateSDKPackage.ps1 will do everything of the starting package.  This chapter is to describe what the script does.
I add a statement 'The script `eng\scripts\automation\Invoke-DataPlaneGenerateSDKPackage.ps1` will do **step-by-step** as following.' above to indicate.

In this section, we'll create a generated API layer built on Azure Core.

From a PowerShell command prompt, navigate to the directory holding `Azure.<group>.<service>.csproj`. Run the following commands:

```
sdk\<your-service-name>\Azure.<group>.<service>\src> dotnet build /t:GenerateCode

> ```

Reviewer 5: Why don't we invoke autorest directly?  This isn't consistent with the way we generate Python. I would like the processes to as similar as possible across our Tier1 languages.

Author: .Net integrate autorest call into dotnet build target. It is the new approach, and go seems accept the same solution.

Reviewer 2: @mikekistler, the autorest call is quite complex - we have this nice tool that simplifies it for us and that our partners can benefit from as well.

Reviewer 5: @annelo-msft Will this tool be implemented to support all the Dataplane Codegen languages? We should not have five different ways of generating DataPlane Codegen libraries.  Ideally the instructions would vary only the parameter passed to autorest to say which language to generate.  I know that's idealistic, but maybe that's the PM's job ;-).

Reviewer 2: @mikekistler, I respectfully disagree that all languages should adopt a twenty(ish) argument command line command in favor of a three word one for the sake of consistency.  Maybe we could think about what the ideal command line would be and find a way for all languages to move toward consistency that way, so we wouldn't have to trade off usability for consistency?

After you run the GenerateCode command, you should find a **Generated** folder in your project. Inside the Generated folder you'll find the ServiceClient and ServiceClientOptions which you can use to interact with the service.

> 

Reviewer 5: Somewhere, and here seems like as good a place as any, we need to tell teams how to add convenience / "grow up" code.  There are certainly right and wrong ways to do this, so we need to give service teams some guidance on this. We could just give the basics here and then point to another document with full details, since most service teams won't bother to add convenience code, but we need the information here in case they do.

Update: I see now something at the very end of the file: **Learn more**.  I think that is way too late and it needs to point to specific material on adding convenience/grow up -- service teams are not going to wade through the whole readme to find this information.

Author: I updated in customize chapter. Please help to review

Reviewer 2: @mikekistler, I think we should wait until after the "starter package" section to provide details regarding customization.  To make sure we get it in sooner, we could summarize the process in the introduction, mention in the starter package that you can do customization after the starter package is generated and checked-in (and passing CI), and then provide details in the customization section.

Reviewer 5: Good point that we probably want to get the generated code checked in and passing CI before adding any extra code.

### Export public API

If you make public API changes or additions, the `eng\scripts\Export-API.ps1` script has to be run to update public API listings. This generates a file in the library's directory similar to the example found in [sdk\template-dpg\Azure.ServiceTemplate.Template\api\Azure.ServiceTemplate.Template.netstandard2.0.cs](https://github.com/Azure/azure-sdk-for-net/blob/bb0fbccfc33dd27d1ec6f0870022824d47181e61/sdk/template-dpg/Azure.ServiceTemplate.Template/api/Azure.ServiceTemplate.Template.netstandard2.0.cs).

e.g. Running the script for a project in `sdk\deviceupdate` would look like this: 
```
eng\scripts\Export-API.ps1 deviceupdate
```

> ## Add package ship requirements

Reviewer 2: I'm wondering if it would make sense to number these steps so it's clear to the reader how the sections relate to each other.

Author: update to use list instead of chapter

> ## Add package ship requirements

Reviewer 2: It would be good to start this section with a statement about what's required by the Azure SDK Guidelines to ship a library.  You can link to those sections of the guidelines as well.

e.g.
In order to ship an Azure SDK library, there are additional things that must be added to the package, including Tests, ... as described in https://azure.github.io/azure-sdk/dotnet_introduction.html#dotnet-repository

Author: updated

> You can continuously work on the first-version shippable package created above to refine, such as add Tests, Samples and refine the Readme.md and Changelog.

Reviewer 2: I believe we moved away from the "shippable" terminology.

```suggestion
You can continuously work on the starting package created above to refine, such as add Tests, Samples and refine the Readme.md and Changelog.
```

Author: updated

> You can continuously work on the first-version shippable package created above to refine, such as add Tests, Samples and refine the Readme.md and Changelog.

Reviewer 2: Let's also say - instead of "continuously work on":

"Before the library package can be released

Please include these links describing what's required for the release process: 
- Tests: https://azure.github.io/azure-sdk/general_implementation.html#testing
- README/Samples/Changelog: https://azure.github.io/azure-sdk/dotnet_introduction.html#dotnet-repository
- Other release concerns: https://azure.github.io/azure-sdk/policies_releases.html

### Tests ###

> 

Reviewer 5: I think we should be sure to set expectations here.  I've added this to the Python DataPlane Codegen quickstart:

> You should write tests that ensure all APIs fulfil their contract and algorithms work as specified.
> This is a requirement of the [Azure SDK Guidelines](https://azure.github.io/azure-sdk/general_implementation.html#testing).

Author: This chapter is to describe how to generate the starter package, and we will ask customer to add Test in 'Add package ship requirements'

> In this section, we will talk about adding unit tests and live tests and how to run them. You will notice that there is a test project under Azure.<group>.<service>\tests.

Reviewer 2: ```suggestion
In this section, we will talk about adding unit tests and live tests and how to run them. You will notice that there is a test project under `Azure.<group>.<service>\tests`.
```

> Here is the step by step process to add tests:

Reviewer 2: @JoshLove-msft, would you be willing to review this section on tests?  This is documentation for our LLC service partners.

- Add other client parameters in `<client-name>ClientTestEnvironment.cs`
- Update `<client-name>ClientTest.cs`. 
  - Comment-out the 'CreateClient' method, and update the new <service>Client statement.

>   - remove all the template project tests, and write the tests according to the commented Test method template. Please refer [this](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#live-testing) to add live tests.

Reviewer 7: It would be better to link to the Test Framework readme - https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core.TestFramework/README.md

Author: The doc as add refer to the 'live test' doc (https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#live-testing), and in the doc, it has a link to the Test Framework readme to tell customer how to write test under Test Framework.

Is it enough? or we should add the TestFramework readme link specifically again in the doc?

**Note**: 

> - Before running live tests you will need to create live test resources, please refer [this](https://github.com/Azure/azure-sdk-for-net/blob/main/eng/common/TestResources/README.md) to learn more about how to manager resources and update test environment.

Reviewer 7: Refer to test framework readme instead - https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core.TestFramework/README.md

Author: updated

> - Before running live tests you will need to create live test resources, please refer [this](https://github.com/Azure/azure-sdk-for-net/blob/main/eng/common/TestResources/README.md) to learn more about how to manager resources and update test environment.

Reviewer 7: ```suggestion
- Before running live tests you will need to create live test resources, please refer to [this](https://github.com/Azure/azure-sdk-for-net/blob/main/eng/common/TestResources/README.md) to learn more about how to manage resources and update test environment.
```

Author: updated

- `Azure.<group>.<service>` is the Azure SDK package name and `<client-name>` is a client name, C# generator will generate a client which you can find in `Azure.<group>.<service>/Generated` directory.

**Learn more:** see the [docs](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#to-test-1) to learn more about tests.

### Samples

> In this section, we will talk about how to add samples. As you can see, we already have a `Samples` folder under `Azure.<group>.<service>/tests` directory. We run the samples as a part of tests. First, enter `<client-name>ClientSamples.cs` and remove the existing commented sample tests. You will add the basic sample tests for your SDK in this file. Create more test files and add tests as per your scenarios.

Reviewer 2: > enter `<client-name>ClientSamples.cs`

This should have the same `Sample<sample_number>_<scenario>.cs` format.

Author: As above comment, we currently cannot know the scenarios of the Client library, we cannot auto create the right file.

Reviewer 2: I'm not suggesting that we know their scenarios, just that we can inform them of the right naming format to use for these files according to the guidelines: https://azure.github.io/azure-sdk/dotnet_introduction.html#dotnet-samples

Author: updated

**Learn more:** For general information about samples, see the [Samples Guidelines](https://azure.github.io/azure-sdk/dotnet_introduction.html#dotnet-samples).

> You will update all the <operation>.md and README.md files under Azure.<group>.<service>\samples directory to the your service according to the examples in those files. Based on that [here](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/anomalydetector/Azure.AI.AnomalyDetector/samples/) is an example.

Reviewer 2: ```suggestion
You will update all the `Sample<sample_number>_<scenario>.md and README.md files under `Azure.<group>.<service>\samples` directory to the your service according to the examples in those files. Based on that [here](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/anomalydetector/Azure.AI.AnomalyDetector/samples/) is an example.
```

One additional thing - I don't think they will update the `Sample<number>_<scenario>.md` files so much as create them to cover the most common use case scenarios.

Author: Currently we have not supported automatically generate the tests and samples, and need manual add. We can not automatically create these files since we have no ideal know what scenario the client contains. Codegen will support to auto-gen those (test, samples, docs, examples), then we will remove the manual work.

Author: updated

**Learn more:** For general information about samples, see the [Samples Guidelines](https://azure.github.io/azure-sdk/dotnet_introduction.html#dotnet-samples).

## Snippets
You will update all the <operation>.md and README.md files under Azure.<group>.<service>\samples directory to the your service according to the examples in those files. Based on that [here](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/anomalydetector/Azure.AI.AnomalyDetector/samples/) is an example.

Snippets are the great way to reuse the sample code. Snippets allow us to verify that the code in our samples and READMEs is always up to date, and passes unit tests. We have added the snippet [here](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/template-dpg/Azure.Template.Generated/tests/Samples/TemplateServiceSamples.HelloWorld.cs#L30) in a sample and used it in the [README](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/template-dpg/Azure.Template.Generated/README.md#create-resource). Please refer [this](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#updating-sample-snippets) to add snippets in your samples.
### Snippets

> ## Export Public API

Reviewer 2: Curious why this section was removed?  It is a requirement, and without doing it the CI will fail.

Author: Move this task into 'starting-package' generation. So starting package already contains API file, No need to manually add.

> In Azure.<group>.<service> project, we have [autorest.md](https://github.com/Azure/azure-sdk-for-net/blob/bb0fbccfc33dd27d1ec6f0870022824d47181e61/sdk/template-dpg/Azure.ServiceTemplate.Template/src/autorest.md) file which is use to add configuration to generate code based on your swagger. After you run the GenerateCode command, you should find a Generated folder in your project. Inside the Generated folder you'll find the <Service>Client and <Service>ClientOptions which you can use to interact with the service.

Reviewer 2: It's worth linking to the guidelines on the Service Client here: https://azure.github.io/azure-sdk/dotnet_introduction.html#dotnet-client

It would be good to say that the generator has generated a service client that follows the official .NET Azure SDK Guidelines.

Author: Service Client is implemented in codegen, so customer does not need to manually customize it. So we will not need to put this scenario here. What's your opinion?

> In Azure.<group>.<service> project, we have [autorest.md](https://github.com/Azure/azure-sdk-for-net/blob/bb0fbccfc33dd27d1ec6f0870022824d47181e61/sdk/template-dpg/Azure.ServiceTemplate.Template/src/autorest.md) file which is use to add configuration to generate code based on your swagger. After you run the GenerateCode command, you should find a Generated folder in your project. Inside the Generated folder you'll find the <Service>Client and <Service>ClientOptions which you can use to interact with the service.

Reviewer 2: Curious: Why is this comment here under "Customize"?  It seems to me that it is talking about generating the code.

It's worth noting that convenience APIs (referenced below in the autorest.csharp README) should be reviewed with the arch board if added.

Author: updated

> In Azure.<group>.<service> project, we have [autorest.md](https://github.com/Azure/azure-sdk-for-net/blob/bb0fbccfc33dd27d1ec6f0870022824d47181e61/sdk/template-dpg/Azure.ServiceTemplate.Template/src/autorest.md) file which is use to add configuration to generate code based on your swagger. After you run the GenerateCode command, you should find a Generated folder in your project. Inside the Generated folder you'll find the <Service>Client and <Service>ClientOptions which you can use to interact with the service.

Reviewer 5: ```suggestion
In `Azure.<group>.<service>` project, we have [autorest.md](https://github.com/Azure/azure-sdk-for-net/blob/bb0fbccfc33dd27d1ec6f0870022824d47181e61/sdk/template-dpg/Azure.ServiceTemplate.Template/src/autorest.md) file which is use to add configuration to generate code based on your swagger. After you run the GenerateCode command, you should find a Generated folder in your project. Inside the Generated folder you'll find the <Service>Client and <Service>ClientOptions which you can use to interact with the service.
```

    - [Tests](#tests)
    - [Samples](#samples)
    - [Snippets](#snippets)
    - [README](#readme)
    - [Changelog](#changelog)
    - [Customize](#customize)

<!-- /TOC -->

## Prerequisites

- Install VS 2020 (Community or higher) and make sure you have the [latest updates](https://www.visualstudio.com/).
  - Need at least .NET Framework 4.6.1 and 4.7 development tools
  - Install the **.NET Core cross-platform development** workloads in VisualStudio
- Install **.NET 6.0 SDK** for your specific platform. (or a higher version)
- Install **.NET core 3.1** for your specific platform.
- Install the latest version of [git](https://git-scm.com/downloads)
- Install [PowerShell](https://docs.microsoft.com/powershell/scripting/install/installing-powershell), version 7 or higher.
- Install [NodeJS](https://nodejs.org/) (14.x.x).

## Setup your repo

- Fork and clone an [azure-sdk-for-net](https://github.com/Azure/azure-sdk-for-net) repo. Follow the instructions in the [.NET CONTRIBUTING.md](https://github.com/Azure/azure-sdk-for-net/issues/12903) to fork and clone the `azure-sdk-for-net` repo.
- Create a branch to work in. 

> ## Create starting package

Reviewer 2: ```suggestion
## Create starter package
```

```

After you run the GenerateCode command, you should find a **Generated** folder in your project. Inside the Generated folder you'll find the ServiceClient and ServiceClientOptions which you can use to interact with the service.

3. Export public API

If you make public API changes or additions, the `eng\scripts\Export-API.ps1` script has to be run to update public API listings. This generates a file in the library's directory similar to the example found in [sdk\template-dpg\Azure.ServiceTemplate.Template\api\Azure.ServiceTemplate.Template.netstandard2.0.cs](https://github.com/Azure/azure-sdk-for-net/blob/bb0fbccfc33dd27d1ec6f0870022824d47181e61/sdk/template-dpg/Azure.ServiceTemplate.Template/api/Azure.ServiceTemplate.Template.netstandard2.0.cs).

e.g. Running the script for a project in `sdk\deviceupdate` would look like this: 
```
eng\scripts\Export-API.ps1 deviceupdate
```

## Add package ship requirements

Before the library package can be released, you will need to add several requirements manually, including tests, samples, README, and CHANGELOG.
You can refer to following guideline to add those requirements:

> - Tests: https://azure.github.io/azure-sdk/general_implementation.html#testing

Reviewer 2: I noticed that there are a number of .md linting errors in this file.  Would you be willing to use the docs-linting and docs-markdown extensions in VSCode to check for valid markdown?

Author: resolved

### Changelog

Update the CHANGELOG.md file which exists in `Azure.<group>.<service>/CHANGELOG.md`. For general information about the changelog, see the [Changelog Guidelines](https://azure.github.io/azure-sdk/policies_releases.html#change-logs).

### Customize

In `Azure.<group>.<service>` project, we have [autorest.md](https://github.com/Azure/azure-sdk-for-net/blob/bb0fbccfc33dd27d1ec6f0870022824d47181e61/sdk/template-dpg/Azure.ServiceTemplate.Template/src/autorest.md) file which is use to add configuration to generate code based on your swagger. 

You can customize your client libray in two ways:
- Customizations can be done as a transform in `autorest.md`. e.g.

```yaml
directive:
  - from: swagger-document
    where: $.parameters.endpoint
    transform: >
      if ($.format === undefined) {
        $.format = "url";
      }
  - from: swagger-document
    where: $.parameters.Endpoint
    transform: >
      if ($.format === undefined) {
        $.format = "url";
      }
```

> - Add convenience APIs.

Reviewer 2: We should put adding convenience APIs first, and say how to choose between the two approaches.

Reviewer 2: I'm not sure we do, but you could post on the internal Language - DotNet channel to ask.

Author: re-arrange the order of customizations. Put 'adding convenience APIs' first.

  
We will use dotnet project template [Azure.ServiceTemplate.Template](https://github.com/Azure/azure-sdk-for-net/blob/3ac301ac6435c818ad7a9946ab1c4023cee236ff/eng/templates) to automatically create the project.

You can run `eng\scripts\automation\Invoke-DataPlaneGenerateSDKPackage.ps1` to generate the starting SDK client library package directly as following:

```powershell
eng/scripts/automation/Invoke-DataPlaneGenerateSDKPackage.ps1 -service <servicename> -namespace Azure.<group>.<service> -sdkPath <sdkrepoRootPath> -inputfiles <inputfilelink> [-securityScope <securityScope>] [-securityHeaderName <securityHeaderName>]
```

e.g.

```powershell
pwsh /home/azure-sdk-for-net/eng/scripts/automation/Invoke-DataPlaneGenerateSDKPackage.ps1 -service sample -namespace Azure.Template.Sample -sdkPath /home/azure-sdk-for-net -inputfiles https://github.com/Azure/azure-rest-api-specs/blob/73a0fa453a93bdbe8885f87b9e4e9fef4f0452d0/specification/webpubsub/data-plane/WebPubSub/stable/2021-10-01/webpubsub.json -securityScope https://sample/.default
```

**Note**:

> - Use one of the following pre-approved namespace groups (<https://azure.github.io/azure-sdk/dotnet_introduction.html#dotnet-namespaces-approved-list>): Azure.AI, Azure.Analytics, Azure.Communication, Azure.Data, Azure.DigitalTwins, Azure.IoT, Azure.Learn, Azure.Media, Azure.Management, Azure.Messaging, Azure.ResourceManager, Azure.Search, Azure.Security, Azure.Storage, Azure.Template, Azure.Identity, Microsoft.Extensions.Azure

Reviewer 2: `Azure.Template` is not an approved namespace.  

Let's not put the list of approved namespaces here unless we have a plan for keeping them up to date when this list changes.

Reviewer 2: Updated.

> - namespace is the shipped package name, it should be `Azure.<group>.<service>`

Reviewer 2: Please use put backticks (`) around anything that is an argument in the command line, e.g. `-namespace`, here and below.

Reviewer 2: Updated.

- Both AADToken and AzureKey authentication are supported. **AzureKey authentication** is specific to each service, but generally it is documented by that service. e.g [AzureKey authentication of Azure Storage](https://docs.microsoft.com/azure/storage/blobs/authorize-data-operations-portal#use-the-account-access-key). **AADToken** is for any service that supports [TokenCredential authentication](https://docs.microsoft.com/dotnet/api/azure.core.tokencredential?view=azure-dotnet). If your service supports AADToken, just set the parameter **securityScope**(the security scope), and if your service supports AzureKey authentication, set parameter **securityHeaderName**( the security header name). You also can provide both if your service supports two authentications.

The script `eng\scripts\automation\Invoke-DataPlaneGenerateSDKPackage.ps1` will do **step-by-step** as following:

> 1. Create the project folder

Reviewer 2: I'm still concerned that people will think they need to do these steps manually.  Do we need to say what the script does as lines of code, or could we summarize it in words and link to the script for people interested in the details?

Reviewer 2: Updated.

### README

README.md file instructions are listed in `Azure.<group>.<service>/README.md` file. Please add/update the README.md file as per your library.

**Learn more:** to understand more about README, see the [README.md](https://github.com/Azure/azure-sdk-for-net/blob/3ac301ac6435c818ad7a9946ab1c4023cee236ff/eng/templates/Azure.ServiceTemplate.Template/README.md). Based on that [here](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/keyvault/Azure.Security.KeyVault.Keys/README.md) is an example.

### Changelog

Update the CHANGELOG.md file which exists in `Azure.<group>.<service>/CHANGELOG.md`. For general information about the changelog, see the [Changelog Guidelines](https://azure.github.io/azure-sdk/policies_releases.html#change-logs).

### Customize

> You may need to customize the client library a little bit to provide customers better experience when they use the library. 

Reviewer 2: We should say when they will need customizations or not.
